File: src/vodoo/crm.py
"""CRM lead/opportunity operations for Vodoo."""

from __future__ import annotations

import datetime as _dt
from typing import Any

from vodoo.base import (
    _get_console,
    _is_simple_output,
    is_structured_output,
    structured_print,
)
from vodoo.generated.crm import STAGE_FIELDS, GeneratedCRMNamespace

_PIPELINE_FIELDS = [
    "id",
    "name",
    "stage_id",
    "expected_revenue",
    "probability",
    "create_date",
    "partner_id",
    "user_id",
    "team_id",
]

# Default staleness thresholds (days) by stage sequence position.
# First stage gets a higher threshold; later stages get stricter ones.
DEFAULT_STALE_THRESHOLDS: dict[str, int] = {
    "_first": 30,
    "_middle": 45,
    "_late": 60,
}


class CRMNamespace(GeneratedCRMNamespace):
    """CRM leads/opportunities namespace."""

    def create(
        self,
        name: str,
        *,
        partner_id: int | None = None,
        expected_revenue: float | None = None,
        stage_id: int | None = None,
        user_id: int | None = None,
        team_id: int | None = None,
        tag_ids: list[int] | None = None,
        lead_type: str = "opportunity",
        **extra_fields: Any,
    ) -> int:
        """Create a CRM lead or opportunity.

        Args:
            name: Lead/opportunity name.
            partner_id: Customer partner ID.
            expected_revenue: Expected revenue amount.
            stage_id: Pipeline stage ID.
            user_id: Salesperson user ID.
            team_id: Sales team ID.
            tag_ids: List of tag IDs to apply.
            lead_type: ``'lead'`` or ``'opportunity'`` (default).
            **extra_fields: Additional fields to set.

        Returns:
            ID of created record.

        """
        values: dict[str, Any] = {**extra_fields, "name": name, "type": lead_type}
        if partner_id is not None:
            values["partner_id"] = partner_id
        if expected_revenue is not None:
            values["expected_revenue"] = expected_revenue
        if stage_id is not None:
            values["stage_id"] = stage_id
        if user_id is not None:
            values["user_id"] = user_id
        if team_id is not None:
            values["team_id"] = team_id
        if tag_ids is not None:
            values["tag_ids"] = [(6, 0, tag_ids)]
        return self._client.create(self._model, values)

    def stages(self, *, team_id: int | None = None) -> list[dict[str, Any]]:
        """List CRM pipeline stages.

        Args:
            team_id: Filter stages by sales team (None = all stages).

        Returns:
            List of stage dictionaries.

        """
        domain: list[Any] = []
        if team_id is not None:
            domain.append(("team_id", "=", team_id))
        return self._client.search_read(
            "crm.stage",
            domain=domain,
            fields=STAGE_FIELDS,
            order="sequence",
        )

    def pipeline(
        self,
        *,
        team: str | None = None,
        user: str | None = None,
    ) -> dict[str, Any]:
        """Fetch pipeline data and return aggregated summary.

        Returns a dict with keys ``team``, ``date``, ``stages`` (list of
        per-stage aggregates), ``totals``, and ``deals`` (raw deal list).
        """
        domain: list[Any] = [("type", "=", "opportunity")]
        if team:
            domain.append(("team_id.name", "ilike", team))
        if user:
            domain.append(("user_id.name", "ilike", user))

        deals = self._client.search_read(
            self._model,
            domain=domain,
            fields=_PIPELINE_FIELDS,
            limit=0,  # all
        )

        stages = self._client.search_read(
            "crm.stage",
            domain=[],
            fields=STAGE_FIELDS,
            order="sequence",
        )

        return build_pipeline_summary(deals, stages, team=team)


# ---------------------------------------------------------------------------
# Pure aggregation (no client needed)
# ---------------------------------------------------------------------------

_TODAY = None  # allow override in tests


def _today() -> _dt.date:
    return _TODAY or _dt.datetime.now(_dt.UTC).date()


def _age_days(create_date: str | None) -> int:
    if not create_date:
        return 0
    try:
        created = _dt.datetime.fromisoformat(create_date).date()
    except (ValueError, TypeError):
        return 0
    return (_today() - created).days


def build_pipeline_summary(
    deals: list[dict[str, Any]],
    stages: list[dict[str, Any]],
    *,
    team: str | None = None,
) -> dict[str, Any]:
    """Build a pipeline summary from raw deals and stages."""
    stage_order = {s["id"]: i for i, s in enumerate(stages)}
    stage_names = {s["id"]: s["name"] for s in stages}

    # Group deals by stage
    by_stage: dict[int, list[dict[str, Any]]] = {}
    for deal in deals:
        sid = deal.get("stage_id")
        if isinstance(sid, list):
            sid = sid[0]
        if sid is None:
            continue
        by_stage.setdefault(sid, []).append(deal)

    stage_summaries: list[dict[str, Any]] = []
    total_deals = 0
    total_revenue = 0.0
    total_weighted = 0.0

    for stage in stages:
        sid = stage["id"]
        stage_deals = by_stage.get(sid, [])
        count = len(stage_deals)
        if count == 0:
            continue

        revenue = sum(d.get("expected_revenue") or 0 for d in stage_deals)
        weighted = sum(
            (d.get("expected_revenue") or 0) * (d.get("probability") or 0) / 100
            for d in stage_deals
        )
        ages = [_age_days(d.get("create_date")) for d in stage_deals]
        avg_age = round(sum(ages) / count) if count else 0
        oldest = max(ages) if ages else 0

        stage_summaries.append(
            {
                "stage_id": sid,
                "name": stage_names.get(sid, f"Stage {sid}"),
                "deals": count,
                "revenue": revenue,
                "weighted": round(weighted, 2),
                "avg_age_days": avg_age,
                "oldest_days": oldest,
            }
        )

        total_deals += count
        total_revenue += revenue
        total_weighted += weighted

    # Attach enriched deal data (for --deals flag)
    enriched_deals: list[dict[str, Any]] = []
    for deal in deals:
        sid = deal.get("stage_id")
        if isinstance(sid, list):
            sid = sid[0]
        partner = deal.get("partner_id")
        partner_name = partner[1] if isinstance(partner, list) else None
        user = deal.get("user_id")
        user_name = user[1] if isinstance(user, list) else None
        enriched_deals.append(
            {
                "id": deal["id"],
                "name": deal.get("name", ""),
                "stage_id": sid,
                "stage_name": stage_names.get(sid, ""),
                "expected_revenue": deal.get("expected_revenue") or 0,
                "probability": deal.get("probability") or 0,
                "age_days": _age_days(deal.get("create_date")),
                "partner": partner_name,
                "user": user_name,
                "stage_order": stage_order.get(sid, 999),
            }
        )
    enriched_deals.sort(key=lambda d: (d["stage_order"], -d["expected_revenue"]))

    return {
        "team": team or "All Teams",
        "date": str(_today()),
        "stages": stage_summaries,
        "totals": {
            "deals": total_deals,
            "revenue": total_revenue,
            "weighted": round(total_weighted, 2),
        },
        "deals": enriched_deals,
    }


# ---------------------------------------------------------------------------
# Health checks
# ---------------------------------------------------------------------------


def compute_health_flags(
    summary: dict[str, Any],
    *,
    stale_thresholds: dict[str, int] | None = None,
) -> list[dict[str, Any]]:
    """Compute health flags for deals in a pipeline summary.

    Returns a sorted list of ``{severity, rule, deal_id, deal_name, detail}`` dicts.
    """
    thresholds = stale_thresholds or DEFAULT_STALE_THRESHOLDS
    first_threshold = thresholds.get("_first", 30)
    middle_threshold = thresholds.get("_middle", 45)
    late_threshold = thresholds.get("_late", 60)

    stages = summary.get("stages", [])
    stage_positions: dict[int, int] = {}
    for i, s in enumerate(stages):
        stage_positions[s["stage_id"]] = i
    num_stages = len(stages)

    severity_order = {"critical": 0, "warning": 1, "info": 2}
    flags: list[dict[str, Any]] = []

    for deal in summary.get("deals", []):
        did = deal["id"]
        dname = deal["name"]
        prob = deal.get("probability", 0)
        rev = deal.get("expected_revenue", 0)
        age = deal.get("age_days", 0)
        partner = deal.get("partner")
        user = deal.get("user")
        sid = deal.get("stage_id")
        pos = stage_positions.get(sid, 0)

        # Critical: 0% probability on open deal
        if prob == 0:
            flags.append(
                {
                    "severity": "critical",
                    "rule": "Zero probability",
                    "deal_id": did,
                    "deal_name": dname,
                    "detail": "Open deal with 0% probability",
                }
            )

        # Warning: missing revenue past first stage
        if rev == 0 and pos > 0:
            flags.append(
                {
                    "severity": "warning",
                    "rule": "Missing revenue",
                    "deal_id": did,
                    "deal_name": dname,
                    "detail": f"No expected revenue in stage '{deal.get('stage_name', '')}'",
                }
            )

        # Staleness by position
        if pos == 0 and age > first_threshold:
            flags.append(
                {
                    "severity": "info",
                    "rule": "Stale deal",
                    "deal_id": did,
                    "deal_name": dname,
                    "detail": f"{age}d in first stage (threshold: {first_threshold}d)",
                }
            )
        elif 0 < pos < num_stages // 2 and age > middle_threshold:
            flags.append(
                {
                    "severity": "warning",
                    "rule": "Stale deal",
                    "deal_id": did,
                    "deal_name": dname,
                    "detail": f"{age}d in '{deal.get('stage_name', '')}'"
                    f" (threshold: {middle_threshold}d)",
                }
            )
        elif pos >= num_stages // 2 and age > late_threshold:
            flags.append(
                {
                    "severity": "warning",
                    "rule": "Stale deal",
                    "deal_id": did,
                    "deal_name": dname,
                    "detail": f"{age}d in '{deal.get('stage_name', '')}'"
                    f" (threshold: {late_threshold}d)",
                }
            )

        # Warning: no partner
        if not partner:
            flags.append(
                {
                    "severity": "warning",
                    "rule": "No partner",
                    "deal_id": did,
                    "deal_name": dname,
                    "detail": "No partner linked",
                }
            )

        # Warning: no salesperson
        if not user:
            flags.append(
                {
                    "severity": "warning",
                    "rule": "No salesperson",
                    "deal_id": did,
                    "deal_name": dname,
                    "detail": "No salesperson assigned",
                }
            )

    flags.sort(key=lambda f: (severity_order.get(f["severity"], 9), f["deal_name"]))
    return flags


# ---------------------------------------------------------------------------
# Display functions
# ---------------------------------------------------------------------------


def display_crm_stages(stages: list[dict[str, Any]]) -> None:
    """Display CRM pipeline stages in a table, TSV, or structured format."""
    if is_structured_output():
        structured_print(stages)
        return

    if _is_simple_output():
        print("id\tname\tsequence\tis_won\tfold")
        for s in stages:
            won = "true" if s.get("is_won") else "false"
            fold = "true" if s.get("fold") else "false"
            print(f"{s['id']}\t{s['name']}\t{s.get('sequence', '')}\t{won}\t{fold}")
    else:
        from rich.table import Table

        console = _get_console()
        table = Table(show_header=True, header_style="bold magenta")
        table.add_column("ID", style="cyan", justify="right")
        table.add_column("Name", style="green")
        table.add_column("Sequence", justify="right")
        table.add_column("Won", justify="center")
        table.add_column("Folded", justify="center")

        for s in stages:
            table.add_row(
                str(s["id"]),
                s["name"],
                str(s.get("sequence", "")),
                "✓" if s.get("is_won") else "",
                "✓" if s.get("fold") else "",
            )

        console.print(table)


def _fmt_currency(value: float) -> str:
    """Format a number as currency-like string."""
    if value >= 1_000_000:
        return f"{value / 1_000_000:,.1f}M"
    if value >= 1_000:
        return f"{value:,.0f}"
    return f"{value:,.0f}"


def _fmt_days(days: int) -> str:
    return f"{days}d"


def display_pipeline(
    summary: dict[str, Any],
    *,
    show_deals: bool = False,
    show_health: bool = False,
    health_flags: list[dict[str, Any]] | None = None,
) -> None:
    """Display pipeline summary in table, TSV, or structured format."""
    if is_structured_output():
        output: dict[str, Any] = {
            "team": summary["team"],
            "date": summary["date"],
            "stages": summary["stages"],
            "totals": summary["totals"],
        }
        if show_deals:
            output["deals"] = summary["deals"]
        if show_health and health_flags:
            output["health"] = health_flags
        structured_print(output)
        return

    if _is_simple_output():
        _display_pipeline_simple(summary, show_deals, show_health, health_flags)
        return

    _display_pipeline_rich(summary, show_deals, show_health, health_flags)


def _display_pipeline_simple(
    summary: dict[str, Any],
    show_deals: bool,
    show_health: bool,
    health_flags: list[dict[str, Any]] | None,
) -> None:
    print(f"# Pipeline: {summary['team']}  {summary['date']}")
    print("stage\tdeals\trevenue\tweighted\tavg_age\toldest")
    for s in summary["stages"]:
        print(
            f"{s['name']}\t{s['deals']}\t{s['revenue']}\t{s['weighted']}"
            f"\t{s['avg_age_days']}\t{s['oldest_days']}"
        )
    t = summary["totals"]
    print(f"TOTAL\t{t['deals']}\t{t['revenue']}\t{t['weighted']}\t\t")

    if show_deals:
        print("\n# Deals")
        print("id\tname\tstage\trevenue\tprobability\tage\tuser")
        for d in summary["deals"]:
            print(
                f"{d['id']}\t{d['name']}\t{d['stage_name']}"
                f"\t{d['expected_revenue']}\t{d['probability']}"
                f"\t{d['age_days']}\t{d.get('user', '')}"
            )

    if show_health and health_flags:
        print("\n# Health")
        print("severity\trule\tdeal_id\tdeal_name\tdetail")
        for f in health_flags:
            print(f"{f['severity']}\t{f['rule']}\t{f['deal_id']}\t{f['deal_name']}\t{f['detail']}")


def _display_pipeline_rich(
    summary: dict[str, Any],
    show_deals: bool,
    show_health: bool,
    health_flags: list[dict[str, Any]] | None,
) -> None:
    from rich.table import Table

    console = _get_console()

    console.print(f"\n[bold]Pipeline: {summary['team']}[/bold]  [dim]{summary['date']}[/dim]\n")

    table = Table(show_header=True, header_style="bold magenta", show_footer=True)
    table.add_column("Stage", style="green", footer_style="bold")
    table.add_column("Deals", justify="right", footer_style="bold")
    table.add_column("Revenue", justify="right", footer_style="bold")
    table.add_column("Weighted", justify="right", footer_style="bold")
    table.add_column("Avg Age", justify="right")
    table.add_column("Oldest", justify="right")

    for s in summary["stages"]:
        table.add_row(
            s["name"],
            str(s["deals"]),
            _fmt_currency(s["revenue"]),
            _fmt_currency(s["weighted"]),
            _fmt_days(s["avg_age_days"]),
            _fmt_days(s["oldest_days"]),
        )

    t = summary["totals"]
    table.columns[0].footer = "Total"
    table.columns[1].footer = str(t["deals"])
    table.columns[2].footer = _fmt_currency(t["revenue"])
    table.columns[3].footer = _fmt_currency(t["weighted"])

    console.print(table)

    # --deals: individual deals under each stage
    if show_deals:
        console.print()
        stage_deals: dict[int, list[dict[str, Any]]] = {}
        for d in summary["deals"]:
            stage_deals.setdefault(d["stage_id"], []).append(d)

        for s in summary["stages"]:
            sid = s["stage_id"]
            deals = stage_deals.get(sid, [])
            console.print(
                f"\n[bold]{s['name']}[/bold]"
                f" [dim]({s['deals']} deals, {_fmt_currency(s['revenue'])})[/dim]"
            )
            for d in deals:
                flags = ""
                if d["expected_revenue"] == 0 and s != summary["stages"][0]:
                    flags += " [yellow]NO-REV[/yellow]"
                if d["probability"] == 0:
                    flags += " [red]0-PROB[/red]"
                user = d.get("user") or "[dim]unassigned[/dim]"
                console.print(
                    f"  [cyan]#{d['id']}[/cyan]  {d['name']:<30}"
                    f"  {_fmt_currency(d['expected_revenue']):>10}"
                    f"  {d['probability']:>3.0f}%"
                    f"  {_fmt_days(d['age_days']):>5}"
                    f"  {user}{flags}"
                )

    # --health: flagged issues
    if show_health and health_flags:
        console.print("\n[bold]Health Flags[/bold]\n")
        htable = Table(show_header=True, header_style="bold")
        htable.add_column("Severity")
        htable.add_column("Rule")
        htable.add_column("Deal")
        htable.add_column("Detail")

        severity_styles = {
            "critical": "bold red",
            "warning": "yellow",
            "info": "dim",
        }
        for f in health_flags:
            style = severity_styles.get(f["severity"], "")
            htable.add_row(
                f"[{style}]{f['severity'].upper()}[/{style}]",
                f["rule"],
                f"#{f['deal_id']} {f['deal_name']}",
                f["detail"],
            )

        console.print(htable)
    elif show_health:
        console.print("\n[green]No health issues found.[/green]")

File: packages/typescript/src/namespaces/crm.ts
import type { OdooClientApi } from "../client-api.js";
import { Cmd } from "../commands.js";
import { decodeRecordDates, formatOdooDate } from "../dates.js";
import { GeneratedCRMNamespace } from "../generated/crm.js";
import type { OdooRecord } from "../types.js";
import type { ListOptions } from "./domain.js";

const PIPELINE_FIELDS = [
  "id",
  "name",
  "stage_id",
  "expected_revenue",
  "probability",
  "create_date",
  "partner_id",
  "user_id",
  "team_id",
] as const;

export interface StaleThresholds {
  _first?: number;
  _middle?: number;
  _late?: number;
}

export const DEFAULT_STALE_THRESHOLDS = Object.freeze({
  _first: 30,
  _middle: 45,
  _late: 60,
});

export interface CRMRecord extends OdooRecord {
  id: number;
  name: string;
  create_date?: Date | null;
}

export interface CRMStage extends OdooRecord {
  id: number;
  name: string;
  sequence?: number;
  is_won?: boolean | null;
  fold?: boolean | null;
}

export interface CreateCRMOptions {
  partnerId?: number;
  expectedRevenue?: number;
  stageId?: number;
  userId?: number;
  teamId?: number;
  tagIds?: readonly number[];
  leadType?: "lead" | "opportunity";
  extraFields?: Readonly<Record<string, unknown>>;
}

export interface PipelineStageSummary {
  stageId: number;
  name: string;
  deals: number;
  revenue: number;
  weighted: number;
  avgAgeDays: number;
  oldestDays: number;
}

export interface PipelineDeal {
  id: number;
  name: string;
  stageId: number | null;
  stageName: string;
  expectedRevenue: number;
  probability: number;
  ageDays: number;
  partner: string | null;
  user: string | null;
  stageOrder: number;
}

export interface PipelineSummary {
  team: string;
  date: string;
  stages: PipelineStageSummary[];
  totals: { deals: number; revenue: number; weighted: number };
  deals: PipelineDeal[];
}

export type HealthSeverity = "critical" | "warning" | "info";
export interface HealthFlag {
  severity: HealthSeverity;
  rule: string;
  dealId: number;
  dealName: string;
  detail: string;
}

function relationId(value: unknown): number | null {
  if (typeof value === "number") return value;
  if (Array.isArray(value) && typeof value[0] === "number") return value[0];
  return null;
}

function relationName(value: unknown): string | null {
  return Array.isArray(value) && typeof value[1] === "string" ? value[1] : null;
}

function numberValue(value: unknown): number {
  return typeof value === "number" ? value : 0;
}

/** Match Python's round-to-nearest, ties-to-even behavior for IEEE-754 doubles. */
function roundHalfEven(value: number, digits = 0): number {
  if (!Number.isFinite(value)) return value;
  if (!Number.isInteger(digits) || digits < 0) {
    throw new RangeError("digits must be a non-negative integer");
  }

  const view = new DataView(new ArrayBuffer(8));
  view.setFloat64(0, Math.abs(value), false);
  const bits = view.getBigUint64(0, false);
  const exponentBits = Number((bits >> 52n) & 0x7ffn);
  const fraction = bits & ((1n << 52n) - 1n);
  if (exponentBits === 0 && fraction === 0n) return value;

  const significand = exponentBits === 0 ? fraction : fraction | (1n << 52n);
  const binaryExponent =
    (exponentBits === 0 ? 1 - 1023 : exponentBits - 1023) - 52;
  const decimalFactor = 10n ** BigInt(digits);
  let numerator = significand * decimalFactor;
  let denominator = 1n;
  if (binaryExponent >= 0) {
    numerator <<= BigInt(binaryExponent);
  } else {
    denominator <<= BigInt(-binaryExponent);
  }

  let quotient = numerator / denominator;
  const doubledRemainder = (numerator % denominator) * 2n;
  if (
    doubledRemainder > denominator ||
    (doubledRemainder === denominator && quotient % 2n !== 0n)
  ) {
    quotient += 1n;
  }
  if (value < 0) quotient = -quotient;
  return Number(quotient) / 10 ** digits;
}

function ageDays(value: unknown, today: Date): number {
  if (!(value instanceof Date) || Number.isNaN(value.getTime())) return 0;
  const start = Date.UTC(
    value.getUTCFullYear(),
    value.getUTCMonth(),
    value.getUTCDate(),
  );
  const end = Date.UTC(
    today.getUTCFullYear(),
    today.getUTCMonth(),
    today.getUTCDate(),
  );
  return Math.floor((end - start) / 86_400_000);
}

export function buildPipelineSummary(
  deals: readonly OdooRecord[],
  stages: readonly OdooRecord[],
  options: { team?: string; today?: Date } = {},
): PipelineSummary {
  const today = options.today ?? new Date();
  const stageOrder = new Map<number, number>();
  const stageNames = new Map<number, string>();
  stages.forEach((stage, index) => {
    if (typeof stage.id === "number") {
      stageOrder.set(stage.id, index);
      stageNames.set(stage.id, String(stage.name ?? `Stage ${stage.id}`));
    }
  });

  const byStage = new Map<number, OdooRecord[]>();
  for (const deal of deals) {
    const stageId = relationId(deal.stage_id);
    if (stageId === null) continue;
    const current = byStage.get(stageId) ?? [];
    current.push(deal);
    byStage.set(stageId, current);
  }

  const stageSummaries: PipelineStageSummary[] = [];
  let totalDeals = 0;
  let totalRevenue = 0;
  let totalWeighted = 0;
  for (const stage of stages) {
    if (typeof stage.id !== "number") continue;
    const stageDeals = byStage.get(stage.id) ?? [];
    if (stageDeals.length === 0) continue;
    const revenue = stageDeals.reduce(
      (total, deal) => total + numberValue(deal.expected_revenue),
      0,
    );
    const weighted = stageDeals.reduce(
      (total, deal) =>
        total +
        (numberValue(deal.expected_revenue) * numberValue(deal.probability)) /
          100,
      0,
    );
    const ages = stageDeals.map((deal) => ageDays(deal.create_date, today));
    stageSummaries.push({
      stageId: stage.id,
      name: stageNames.get(stage.id) ?? `Stage ${stage.id}`,
      deals: stageDeals.length,
      revenue,
      weighted: roundHalfEven(weighted, 2),
      avgAgeDays: roundHalfEven(
        ages.reduce((total, age) => total + age, 0) / stageDeals.length,
      ),
      oldestDays: Math.max(...ages),
    });
    totalDeals += stageDeals.length;
    totalRevenue += revenue;
    totalWeighted += weighted;
  }

  const enrichedDeals: PipelineDeal[] = deals
    .filter((deal): deal is OdooRecord & { id: number } =>
      Number.isInteger(deal.id),
    )
    .map((deal) => {
      const stageId = relationId(deal.stage_id);
      return {
        id: deal.id,
        name: String(deal.name ?? ""),
        stageId,
        stageName: stageId === null ? "" : (stageNames.get(stageId) ?? ""),
        expectedRevenue: numberValue(deal.expected_revenue),
        probability: numberValue(deal.probability),
        ageDays: ageDays(deal.create_date, today),
        partner: relationName(deal.partner_id),
        user: relationName(deal.user_id),
        stageOrder: stageId === null ? 999 : (stageOrder.get(stageId) ?? 999),
      };
    })
    .sort(
      (left, right) =>
        left.stageOrder - right.stageOrder ||
        right.expectedRevenue - left.expectedRevenue,
    );

  return {
    team:
      options.team === undefined || options.team === ""
        ? "All Teams"
        : options.team,
    date: formatOdooDate(today),
    stages: stageSummaries,
    totals: {
      deals: totalDeals,
      revenue: totalRevenue,
      weighted: roundHalfEven(totalWeighted, 2),
    },
    deals: enrichedDeals,
  };
}

export function computeHealthFlags(
  summary: PipelineSummary,
  thresholds: Readonly<StaleThresholds> = DEFAULT_STALE_THRESHOLDS,
): HealthFlag[] {
  const firstThreshold = thresholds._first ?? 30;
  const middleThreshold = thresholds._middle ?? 45;
  const lateThreshold = thresholds._late ?? 60;
  const stagePositions = new Map(
    summary.stages.map((stage, index) => [stage.stageId, index]),
  );
  const severityOrder: Record<HealthSeverity, number> = {
    critical: 0,
    warning: 1,
    info: 2,
  };
  const flags: HealthFlag[] = [];
  for (const deal of summary.deals) {
    const position =
      deal.stageId === null ? 0 : (stagePositions.get(deal.stageId) ?? 0);
    if (deal.probability === 0) {
      flags.push({
        severity: "critical",
        rule: "Zero probability",
        dealId: deal.id,
        dealName: deal.name,
        detail: "Open deal with 0% probability",
      });
    }
    if (deal.expectedRevenue === 0 && position > 0) {
      flags.push({
        severity: "warning",
        rule: "Missing revenue",
        dealId: deal.id,
        dealName: deal.name,
        detail: `No expected revenue in stage '${deal.stageName}'`,
      });
    }
    const halfway = Math.floor(summary.stages.length / 2);
    const staleThreshold =
      position === 0
        ? firstThreshold
        : position < halfway
          ? middleThreshold
          : lateThreshold;
    if (deal.ageDays > staleThreshold) {
      flags.push({
        severity: position === 0 ? "info" : "warning",
        rule: "Stale deal",
        dealId: deal.id,
        dealName: deal.name,
        detail:
          position === 0
            ? `${deal.ageDays}d in first stage (threshold: ${staleThreshold}d)`
            : `${deal.ageDays}d in '${deal.stageName}' (threshold: ${staleThreshold}d)`,
      });
    }
    if (deal.partner === null) {
      flags.push({
        severity: "warning",
        rule: "No partner",
        dealId: deal.id,
        dealName: deal.name,
        detail: "No partner linked",
      });
    }
    if (deal.user === null) {
      flags.push({
        severity: "warning",
        rule: "No salesperson",
        dealId: deal.id,
        dealName: deal.name,
        detail: "No salesperson assigned",
      });
    }
  }
  return flags.sort(
    (left, right) =>
      severityOrder[left.severity] - severityOrder[right.severity] ||
      left.dealName.localeCompare(right.dealName),
  );
}

export class CRMNamespace extends GeneratedCRMNamespace {
  constructor(client: OdooClientApi) {
    super(client);
  }

  override async list(options: ListOptions = {}): Promise<CRMRecord[]> {
    return (await super.list(options)) as CRMRecord[];
  }

  override async get(
    recordId: number,
    fields?: readonly string[],
  ): Promise<CRMRecord> {
    return (await super.get(recordId, fields)) as CRMRecord;
  }

  create(name: string, options: CreateCRMOptions = {}): Promise<number> {
    const values: Record<string, unknown> = {
      ...options.extraFields,
      name,
      type: options.leadType ?? "opportunity",
    };
    if (options.partnerId !== undefined) values.partner_id = options.partnerId;
    if (options.expectedRevenue !== undefined) {
      values.expected_revenue = options.expectedRevenue;
    }
    if (options.stageId !== undefined) values.stage_id = options.stageId;
    if (options.userId !== undefined) values.user_id = options.userId;
    if (options.teamId !== undefined) values.team_id = options.teamId;
    if (options.tagIds !== undefined)
      values.tag_ids = [Cmd.set(options.tagIds)];
    return this.client.create("crm.lead", values);
  }

  override async stages(teamId: number | null = null): Promise<CRMStage[]> {
    return (await super.stages(teamId)) as CRMStage[];
  }

  async pipeline(
    options: { team?: string; user?: string } = {},
  ): Promise<PipelineSummary> {
    const domain: Array<readonly [string, string, unknown]> = [
      ["type", "=", "opportunity"],
    ];
    if (options.team !== undefined && options.team !== "") {
      domain.push(["team_id.name", "ilike", options.team]);
    }
    if (options.user !== undefined && options.user !== "") {
      domain.push(["user_id.name", "ilike", options.user]);
    }
    const deals = await this.client.searchRead("crm.lead", {
      domain,
      fields: PIPELINE_FIELDS,
      limit: 0,
    });
    const decodedDeals = deals.map((deal) =>
      decodeRecordDates(deal, { create_date: "datetime" }),
    );
    const stages = await this.client.searchRead("crm.stage", {
      domain: [],
      fields: ["id", "name", "sequence", "is_won", "fold"],
      order: "sequence",
    });
    const summaryOptions =
      options.team === undefined ? {} : { team: options.team };
    return buildPipelineSummary(decodedDeals, stages, summaryOptions);
  }
}
